
# DL Assignment 03

**Name:Md. Asad-Al-Adil**

**Course Email:adilasadals@gmail.com**  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [ ]:
# library import
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder,StandardScaler

In [ ]:
df=pd.read_csv("https://raw.githubusercontent.com/asadaladil/Others-Materials/refs/heads/main/insurance.csv")
print(df.shape)
print(df.columns)

target_variable='smoker'

df.head(10)

(1338, 7)
Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520
5,31,female,25.740,0,no,southeast,3756.62160
6,46,female,33.440,1,no,southeast,8240.58960
7,37,female,27.740,3,no,northwest,7281.50560
8,37,male,29.830,2,no,northeast,6406.41070
9,60,female,25.840,0,no,northwest,28923.13692


In [ ]:
print(df.isnull().sum())
df.describe()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64


,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


In [ ]:
numerical_col=['age','bmi','children','charges']
categorical_col=['sex','smoker','region']

df_encoded=pd.get_dummies(df, columns=categorical_col, dtype=int,drop_first=True)
temp=df_encoded['smoker_yes']
df_encoded.drop('smoker_yes',axis=1,inplace=True)
df_encoded['smoker']=temp

df_encoded.head()

,age,bmi,children,charges,sex_male,region_northwest,region_southeast,region_southwest,smoker
0,19,27.900,0,16884.92400,0,0,0,1,1
1,18,33.770,1,1725.55230,1,0,1,0,0
2,28,33.000,3,4449.46200,1,0,1,0,0
3,33,22.705,0,21984.47061,1,1,0,0,0
4,32,28.880,0,3866.85520,1,1,0,0,0


In [ ]:
X=df_encoded.drop(columns=[target_variable])
y=df_encoded[target_variable]

scaler = StandardScaler()

X_numpy=scaler.fit_transform(X)
y_numpy=y.to_numpy()

torch.manual_seed(100)
X_torch=torch.tensor(X_numpy,dtype=torch.float32)
y_torch=torch.tensor(y_numpy,dtype=torch.float32).view(-1,1)
X_train,X_test,y_train,y_test=train_test_split(X_torch,y_torch,test_size=0.2,random_state=42)
print(X_train.shape)

torch.Size([1070, 8])


##Explaination:
1. Handling Missing Values

- The Problem: Neural networks and mathematical models cannot perform calculations on NaN or Null values. So, we need to remove or fill the null value.

2. Encoding Categorical Variables
- Neural networks only understand numbers, not text like "Red", "Green", or "Blue". We use one-hot encoding to convert categories into numerical value.


3. Separating Features (X) and Target (y)
- In supervised learning, the model needs to learn the relationship between inputs and the desired output.


4. Splitting into Train and Test Sets
- If we train a model on all the data, it might just memorize the answers & it will cause overfitting. We won't know if it can actually generalize to new, unseen data.


5. Feature Scaling (StandardScaler)
- We scale all numerical features to have a mean of 0 and a standard deviation of 1. This puts all features on an equal playing field so the neural network optimizes smoothly.


6. Converting to NumPy Arrays
- Pandas DataFrames are great for data manipulation, but they carry a lot of overhead like indexes and column names.


7. Converting to PyTorch Tensors
- PyTorch cannot use NumPy arrays directly for training. It requires its own data structure called a Tensor, which is optimized to run massively parallel calculations on a GPU.


# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [ ]:
class Model(nn.Module):
  def __init__(self,input_size,hidden_size,output_size):
    super().__init__()
    self.linear1=nn.Linear(input_size,hidden_size)
    self.relu=nn.ReLU()
    self.linear2=nn.Linear(hidden_size,output_size)
    self.sigmoid=nn.Sigmoid()

  def forward(self,X):
    out=self.linear1(X)
    out=self.relu(out)
    out=self.linear2(out)
    out=self.sigmoid(out)
    return out

In [ ]:
model=Model(X_torch.shape[1],32,1)
model(X_torch)

tensor([[0.4017],
        [0.3565],
        [0.4121],
        ...,
        [0.3911],
        [0.3723],
        [0.3714]], grad_fn=<SigmoidBackward0>)

##Justification:
Number of Hidden Neurons:
- The number of hidden neurons controls the learning capacity of the model. If hidden_size is too small, the model suffers from underfitting and cannot learn complex patterns in the data. If it is too large, the model suffers from overfitting, where it memorizes the training data but fails to generalize to new, unseen data. It also makes training slower.


Choice of Activation Functions:
- I used two different activation functions here.

- ReLU (Hidden Layer): The Rectified Linear Unit is the industry standard for hidden layers. It introduces the necessary non-linearity so the network can learn complex, curved boundaries rather than just straight lines. It is also computationally cheap and prevents the vanishing gradient problem, allowing the model to learn quickly during backpropagation.

- Sigmoid (Output Layer): The Sigmoid function mathematically squashes whatever raw number the network outputs into a strict range between 0.0 and 1.0. This allows the model to interpret the final output as a probability.

Code to Print Trainable Parameters:
- To print the exact number of parameters, we have to instantiate the model with real numbers. Here is how you calculate and print the total parameters that PyTorch will be tweaking during training:

- Where that number comes from (8,32,1):

- linear1: (8 inputs * 32 neurons) + 32 biases = 256 parameters

- linear2: (32 inputs * 1 neuron) + 1 bias = 33 parameters

- Total: 256 + 33 = 289 parameters.

# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

- Loss function: Binary Cross Entropy(BCE)
- Optimizer: Stochastic Gradient Descent (SGD):

Justification (Regression vs. Classification):

- The model ends with a nn.Sigmoid() layer, which results the output into a probability between 0.0 and 1.0. Because the model is predicting discrete categories rather than a continuous number, this is a binary classification task. BCE is the mathematically optimal loss function for this. It calculates the logarithmic distance between your model's predicted probability and the actual true label (0 or 1).

- If this were a regression task like predicting a house's price in dollars, I would use Mean Squared Error (nn.MSELoss). Using MSE for classification is a poor choice because it doesn't heavily penalize the model for being entirely wrong but highly confident. BCE applies a massive logarithmic penalty when the model predicts 0.99 for a label that is actually 0, forcing the network to learn rapidly from its worst mistakes.

Justification (Nature of the dataset):

- Standard datasets often contain noise or outliers. SGD updates the model's weights using only a small subset of the data at a time, it introduces a bit of noise into the training path. This noise actually acts as a good natural regularizer, preventing the model from memorizing the training data and helping it generalize better to unseen testing data.


In [ ]:
loss_function=nn.BCELoss()
optimizer=torch.optim.SGD(model.parameters(), lr=0.01)

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [ ]:
epochs=100
loss_history=[]

for epoch in range(epochs):
    y_pred=model.forward(X_train)
    #print(y_pred.shape,y_train.shape)
    loss=loss_function(y_pred,y_train)
    loss.backward()
    optimizer.step()
    loss_history.append(loss.item())
    if (epoch+1)%10==0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

    optimizer.zero_grad()

print("Training complete!")

Epoch [10/100], Loss: 0.6189
Epoch [20/100], Loss: 0.6028
Epoch [30/100], Loss: 0.5884
Epoch [40/100], Loss: 0.5756
Epoch [50/100], Loss: 0.5641
Epoch [60/100], Loss: 0.5537
Epoch [70/100], Loss: 0.5442
Epoch [80/100], Loss: 0.5357
Epoch [90/100], Loss: 0.5278
Epoch [100/100], Loss: 0.5206
Training complete!


##Pipeline:

Forward Pass:
- Input input data is pushed through the network. It gets multiplied by the weights in linear1, passes through the ReLU activation, goes through linear2, and finally gets squashed by the Sigmoid activation. The output is the model's current "best guess" (a probability between 0.0 and 1.0).

Loss Computation
- The BCELoss function takes the model's predictions and compares them directly to the true labels. It outputs a single scalar value representing how "wrong" the model was on this batch of data.

Backward Pass
- PyTorch's autograd engine performs backpropagation. It uses calculus to travel backward through the network, calculating the gradient for every single weight and bias. These gradients tell the optimizer exactly which direction to adjust each parameter to lower the error.

Parameter Update
- The SGD optimizer steps in, takes the gradients calculated in the previous step, and applies them to the network's weights. It updates the parameters by taking a small step down the gradient slope.

Gradient Reset
- As PyTorch accumulates gradients, it means if we don't explicitly clear the gradients from the previous epoch, the new gradients will just be added on top of the old ones. This will completely ruin the math and prevent the model from learning. We call this first to ensure we are starting with a clean slate.

# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [ ]:
# for classification:

#Train Performance
y_pred=model.forward(X_train)
y_pred=(y_pred>=0.5).float()
accuracy=(y_pred==y_train).float().mean()
print(f'Train Accuracy: {accuracy.item():.4f}')


#Test Performance
y_pred=model.forward(X_test)
y_pred=(y_pred>=0.5).float()
accuracy=(y_pred==y_test).float().mean()
print(f'Test Accuracy: {accuracy.item():.4f}')


Train Accuracy: 0.7944
Test Accuracy: 0.7985


Overfitting Problem:
- Overfitting happens when a model memorizes the training data but fails to generalize to new data. If the model were overfitting, their would see a Training Accuracy in the 90s and a Testing Accuracy much lower like 70%.

Because the Test Accuracy (79.85%) is virtually identical to your Train Accuracy (79.44%)—in fact, it is slightly higher—your model is generalizing perfectly to unseen data. It is successfully applying the rules it learned rather than just memorizing answers.

Underfitting Problem:
- This is a healthy learning curve. However, whether it is mildly underfitting depends entirely on your dataset:

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [ ]:
model=Model(X_torch.shape[1],128,1)
optimizer=torch.optim.SGD(model.parameters(), lr=0.05)

epochs=1000
loss_history=[]

for epoch in range(epochs):
    y_pred=model.forward(X_train)
    loss=loss_function(y_pred,y_train)
    loss.backward()
    optimizer.step()
    loss_history.append(loss.item())
    if (epoch+1)%10==0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

    optimizer.zero_grad()

print("Training complete!")

Epoch [10/1000], Loss: 0.4873
Epoch [20/1000], Loss: 0.4133
Epoch [30/1000], Loss: 0.3638
Epoch [40/1000], Loss: 0.3265
Epoch [50/1000], Loss: 0.2972
Epoch [60/1000], Loss: 0.2736
Epoch [70/1000], Loss: 0.2544
Epoch [80/1000], Loss: 0.2384
Epoch [90/1000], Loss: 0.2250
Epoch [100/1000], Loss: 0.2136
Epoch [110/1000], Loss: 0.2039
Epoch [120/1000], Loss: 0.1955
Epoch [130/1000], Loss: 0.1881
Epoch [140/1000], Loss: 0.1817
Epoch [150/1000], Loss: 0.1761
Epoch [160/1000], Loss: 0.1711
Epoch [170/1000], Loss: 0.1667
Epoch [180/1000], Loss: 0.1627
Epoch [190/1000], Loss: 0.1591
Epoch [200/1000], Loss: 0.1559
Epoch [210/1000], Loss: 0.1530
Epoch [220/1000], Loss: 0.1504
Epoch [230/1000], Loss: 0.1479
Epoch [240/1000], Loss: 0.1457
Epoch [250/1000], Loss: 0.1437
Epoch [260/1000], Loss: 0.1418
Epoch [270/1000], Loss: 0.1401
Epoch [280/1000], Loss: 0.1384
Epoch [290/1000], Loss: 0.1369
Epoch [300/1000], Loss: 0.1355
Epoch [310/1000], Loss: 0.1342
Epoch [320/1000], Loss: 0.1330
Epoch [330/1000],

In [ ]:
#Train Performance
y_pred=model.forward(X_train)
y_pred=(y_pred>0.5).float()
accuracy=(y_pred==y_train).float().mean()
print(f'Train Accuracy: {accuracy.item():.4f}')


#Test Performance
y_pred=model.forward(X_test)
y_pred=(y_pred>0.5).float()
accuracy=(y_pred==y_test).float().mean()
print(f'Test Accuracy: {accuracy.item():.4f}')

Train Accuracy: 0.9692
Test Accuracy: 0.9627


Final Performance
- Final performance updated from ~79% to an outstanding ~96%. More importantly, the gap between your Training Accuracy (96.92%) and Testing Accuracy (96.27%) is incredibly small—only about 0.65%. This indicates that the model is performing exceptionally well and successfully applying its learned logic to completely new data.

Convergence Speed
- While training for 1000 epochs took more absolute time, the model was able to converge to a much deeper, highly accurate mathematical solution. Because I have given the model a massive increase in parameters (128 neurons), the loss likely dropped much more steadily and reached a significantly lower final value than the previous run.

How the Change Affected the Model (Explanation)
- It has effectively solved an underfitting problem by increasing the model's capacity.
- In the first run, the model simply did not have enough dimensions to capture the complex relationships in the dataset, causing it to converge at 80%.
- By increasing the hidden layer to 128 neurons, it has given the network a much larger workspace to draw highly intricate, non-linear decision boundaries.
- Furthermore, by increasing the epochs to 1000, this has given the optimizer enough time to properly tune those newly added neurons.

Diagnosis:
- It is not overfitting because the test accuracy closely matches the train accuracy. If it were overfitting due to the large 128-neuron layer, the train accuracy would be 99%, but the test accuracy would have crashed down to 70% or 80%.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

Gradients be reset every epoch:
- PyTorch accumulates gradients. When calling loss.backward(), it doesn't replace the old gradients; it adds the newly calculated gradients.

If the learning rate is too high:
- The learning rate determines the step size for the optimizer takes down or the slope of the loss function.
- If the step is too high, the optimizer will step completely over the lowest point of the error valley and land higher up on the opposite side.
- It will diverge from one side of another side.


If the learning rate is too small:
- The optimizer takes little steps. The loss will go down, but it will take an eternity. For standard learning rate it would take 100 epochs might now take 50,000 epochs for slow learning rate.
- Small learning rate does not have proper power to reach out the local or global minimum point of error.

Define layers inside the constructor (__init__) and not inside forward():
- __init__ is for creating the object of a class.
- When defining nn.Linear() inside the constructor, PyTorch allocates memory for the weights and biases of that layer exactly once. These parameters stay in memory.

- forward() is just the roadmap.
- This function executes every single time passing a batch of data into the model. If I define nn.Linear() inside forward(), I am telling PyTorch to completely modify the old layer and create a new one with randomized, untrained weights every single time a new batch of data arrives.

In [ ]:
import torch
w=torch.tensor(2.0,requires_grad=True)
l1=w**2
l1.backward()
print(w.grad)
l2=3*w
l2.backward()
print(w.grad)

tensor(4.)
tensor(7.)


In [ ]:
import torch
w=torch.tensor(2.0,requires_grad=True)
l1=w**2
l1.backward()
print(w.grad)
w.grad.zero_()
l2=3*w
l2.backward()
print(w.grad)

tensor(4.)
tensor(3.)


In [ ]:
# high learning rate:

model=Model(X_torch.shape[1],128,1)
optimizer=torch.optim.SGD(model.parameters(), lr=100)

epochs=1000
loss_history=[]

for epoch in range(epochs):
    y_pred=model.forward(X_train)
    loss=loss_function(y_pred,y_train)
    loss.backward()
    optimizer.step()
    loss_history.append(loss.item())
    if (epoch+1)%10==0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

    optimizer.zero_grad()

print("Training complete!")

Epoch [10/1000], Loss: 20.5607
Epoch [20/1000], Loss: 20.5607
Epoch [30/1000], Loss: 20.5607
Epoch [40/1000], Loss: 20.5607
Epoch [50/1000], Loss: 20.5607
Epoch [60/1000], Loss: 20.5607
Epoch [70/1000], Loss: 20.5607
Epoch [80/1000], Loss: 20.5607
Epoch [90/1000], Loss: 20.5607
Epoch [100/1000], Loss: 20.5607
Epoch [110/1000], Loss: 20.5607
Epoch [120/1000], Loss: 20.5607
Epoch [130/1000], Loss: 20.5607
Epoch [140/1000], Loss: 20.5607
Epoch [150/1000], Loss: 20.5607
Epoch [160/1000], Loss: 20.5607
Epoch [170/1000], Loss: 20.5607
Epoch [180/1000], Loss: 20.5607
Epoch [190/1000], Loss: 20.5607
Epoch [200/1000], Loss: 20.5607
Epoch [210/1000], Loss: 20.5607
Epoch [220/1000], Loss: 20.5607
Epoch [230/1000], Loss: 20.5607
Epoch [240/1000], Loss: 20.5607
Epoch [250/1000], Loss: 20.5607
Epoch [260/1000], Loss: 20.5607
Epoch [270/1000], Loss: 20.5607
Epoch [280/1000], Loss: 20.5607
Epoch [290/1000], Loss: 20.5607
Epoch [300/1000], Loss: 20.5607
Epoch [310/1000], Loss: 20.5607
Epoch [320/1000],

In [ ]:
#Slow learning rate

model=Model(X_torch.shape[1],128,1)
optimizer=torch.optim.SGD(model.parameters(), lr=0.0001)

epochs=1000
loss_history=[]

for epoch in range(epochs):
    y_pred=model.forward(X_train)
    loss=loss_function(y_pred,y_train)
    loss.backward()
    optimizer.step()
    loss_history.append(loss.item())
    if (epoch+1)%10==0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

    optimizer.zero_grad()

print("Training complete!")

Epoch [10/1000], Loss: 0.7232
Epoch [20/1000], Loss: 0.7223
Epoch [30/1000], Loss: 0.7215
Epoch [40/1000], Loss: 0.7206
Epoch [50/1000], Loss: 0.7198
Epoch [60/1000], Loss: 0.7190
Epoch [70/1000], Loss: 0.7181
Epoch [80/1000], Loss: 0.7173
Epoch [90/1000], Loss: 0.7165
Epoch [100/1000], Loss: 0.7156
Epoch [110/1000], Loss: 0.7148
Epoch [120/1000], Loss: 0.7140
Epoch [130/1000], Loss: 0.7132
Epoch [140/1000], Loss: 0.7124
Epoch [150/1000], Loss: 0.7116
Epoch [160/1000], Loss: 0.7108
Epoch [170/1000], Loss: 0.7100
Epoch [180/1000], Loss: 0.7092
Epoch [190/1000], Loss: 0.7084
Epoch [200/1000], Loss: 0.7076
Epoch [210/1000], Loss: 0.7068
Epoch [220/1000], Loss: 0.7060
Epoch [230/1000], Loss: 0.7052
Epoch [240/1000], Loss: 0.7044
Epoch [250/1000], Loss: 0.7037
Epoch [260/1000], Loss: 0.7029
Epoch [270/1000], Loss: 0.7021
Epoch [280/1000], Loss: 0.7013
Epoch [290/1000], Loss: 0.7006
Epoch [300/1000], Loss: 0.6998
Epoch [310/1000], Loss: 0.6991
Epoch [320/1000], Loss: 0.6983
Epoch [330/1000],